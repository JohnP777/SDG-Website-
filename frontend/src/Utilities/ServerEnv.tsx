export const getUrl = () => {
    if (process.env.REACT_APP_BACKEND_URL) {
        return process.env.REACT_APP_BACKEND_URL;
    }
    return 'http://127.0.0.1:8000/'
}