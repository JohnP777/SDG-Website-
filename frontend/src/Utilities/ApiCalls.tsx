import { getUrl } from "./ServerEnv";

export const apiCallPost = async (path: string, body: any, isAuthed: boolean ) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallGet = async (path: string, isAuthed: boolean, headersExtra?: any) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'GET',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : '',
        ...headersExtra
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPatch = async (path: string, body: object, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PATCH',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallDelete = async (path: string, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'DELETE',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPut = async (path: string, body: any, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PUT',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    });
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error);
  }
};
