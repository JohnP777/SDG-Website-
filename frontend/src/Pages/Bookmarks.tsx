import React, { useEffect } from 'react';
import Page from '../Components/Page';
import { Box, Typography } from '@mui/material';
import { useState } from 'react';
import { TextField } from '@mui/material';
import SearchIcon from '@mui/icons-material/Search';
import FilterAltIcon from '@mui/icons-material/FilterAlt';
import { IconButton } from '@mui/material';
import FormControlLabel from '@mui/material/FormControlLabel';
import Checkbox from '@mui/material/Checkbox';
import Bookmark from '../Components/Bookmark';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import { apiCallGet, apiCallPost } from '../Utilities/ApiCalls';
import SearchSDGPlanModal from '../Components/SearchSDGPlanModal';

// Filters out bookmarks based on the filter
const handleBookmarks = (currentSearch: string, bookmarks: any[] = [], showActionBookmarks: boolean, showEducationBookmarks: boolean) => {
  let bookmarksList = bookmarks
                        .filter(suggestion => (suggestion.title || suggestion.actions).toLowerCase().includes(currentSearch.toLowerCase()))

  if (!showActionBookmarks) {
    bookmarksList = bookmarksList.filter(suggestion => (suggestion.type !== 'action'));
  }
  
  if (!showEducationBookmarks) {
    bookmarksList = bookmarksList.filter(suggestion => (suggestion.type !== 'education'));
  }

  return bookmarksList;

}

const Bookmarks = () => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };
  const [currentSearch, setCurrentSearch] = useState('');

  const [bookmarks, setBookmarks] = useState<any[]>([]);
  const [selectedBookmark, setSelectedBookmark] = useState<any | null>(null);

  const [showActionBookmarks, setShowActionBookmarks] = useState(true);
  const [showEducationBookmarks, setShowEducationBookmarks] = useState(true);

  useEffect(() => {
    const fetchBookmarks = async () => {
      // Putting in bookmarks as data so it's not empty
      await Promise.all([
        await apiCallPost('api/auth/bookmark/set/', { bookmark: 'education-2' }, true),
        await apiCallPost('api/auth/bookmark/set/', { bookmark: 'education-3' }, true),
        await apiCallPost('api/auth/bookmark/set/', { bookmark: 'action-6' }, true),
        await apiCallPost('api/auth/bookmark/set/', { bookmark: 'action-7' }, true)
      ]);

      // Gets title of all bookmarks the user has
      const data = await apiCallGet('api/auth/bookmark/get/', true);

      const userBookmarks: any[] = [];

      for (const bookmark of data.bookmarks) {
        const [type, id] = bookmark.split('-');

        if (type === 'action') {
          const actionBookmark = await apiCallGet(`api/sdg-actions/retrieve?q=${encodeURIComponent(id)}`, false);
          if (typeof actionBookmark.sdgs === 'number') {
            actionBookmark.sdgs = [String(actionBookmark.sdgs)];
          }
          actionBookmark.type = 'action';
          userBookmarks.push(actionBookmark);
        } else if (type === 'education') {
          const educationBookmark = await apiCallGet(`api/sdg-education/retrieve?q=${encodeURIComponent(id)}`, false);
          if (typeof educationBookmark.sdgs_related === 'number') {
            educationBookmark.sdgs_related = [educationBookmark.sdgs_related];
          }
          educationBookmark.type = 'education';
          userBookmarks.push(educationBookmark);
        }
      }
      // userBookmarks has the info associated with all the bookmark titles
      setBookmarks(userBookmarks);
      
    }
    fetchBookmarks();
  }, [])

  // Removes bookmark if you press the red bookmark button
  const handleDelete = async (id: string, type: string) => {
    const deleteBookmark = type + '-' + id

    const data = await apiCallPost('api/auth/bookmark/unset/', { bookmark: deleteBookmark }, true);

    if (data.statusCode === 200) {
      setBookmarks(prev => prev.filter(b => !(b.id === id && b.type === type)));
    } else {
      console.log(data)
    }
  }

  return (
    <Page>
      <Box sx={{
        display: 'flex',
        justifyContent: 'center',
        width: '100%',
        flexDirection: 'column'
      }}>
        <Typography variant='h2' sx={{ fontSize: '30px', fontWeight: 'bold', paddingBottom: '20px' }}>Bookmarks</Typography>

        <TextField
          placeholder='Search your bookmarks'
          fullWidth
          onChange={(e) => setCurrentSearch(e.target.value)}
          sx={{
            backgroundColor: '#FFFFFF',
            padding: '5px',
            fontSize: '18px',
            borderRadius: '5px',
            boxSizing: 'border-box',
            '& fieldset': { border: 'none' },
            marginBottom: '20px',
            width: '50%'
          }}
          InputProps={{
            startAdornment:(
              <SearchIcon sx={{ marginRight: '10px' }} />
              ),
            endAdornment:(
              <IconButton aria-label="filter" onClick={handleClick}>
                <FilterAltIcon />
              </IconButton>
            )
          }}
        />
        <Menu
          id='basic-menu'
          anchorEl={anchorEl}
          open={open}
          onClose={handleClose}
          MenuListProps={{
            'aria-labelledby': 'basic-button',
          }}
        >
          <Typography pl={2}>Filter</Typography>
          <MenuItem>
            <FormControlLabel 
              control={<Checkbox checked={showActionBookmarks} onChange={(e) => setShowActionBookmarks(e.target.checked)} />} 
              label='Actions' 
            />
          </MenuItem>
          <MenuItem>
            <FormControlLabel 
              control={<Checkbox checked={showEducationBookmarks} onChange={(e) => setShowEducationBookmarks(e.target.checked)} />} 
              label='Education' 
            />
          </MenuItem>
        </Menu>

        <Box sx={{
          display: 'flex',
          gap: '10px',
          flexWrap: 'wrap'
        }}>
          {
            handleBookmarks(currentSearch, bookmarks, showActionBookmarks, showEducationBookmarks).map((bookmark, index) => (
              <Bookmark 
                key={index} 
                bookmarkTitle={bookmark.title || bookmark.actions}  
                bookmarkType={bookmark.organization || 'Individual'} 
                sdgGoalNumbers={(bookmark.sdgs || bookmark.sdgs_related)}
                onClick={() => setSelectedBookmark(bookmark)}
                onDelete={() => handleDelete(bookmark.id, bookmark.type)}
              />
            ))
          }
        </Box>
      </Box>
      {selectedBookmark && (
        <SearchSDGPlanModal
          title={selectedBookmark.title || selectedBookmark.actions}
          aims={selectedBookmark.aims}
          descriptions={selectedBookmark.descriptions || selectedBookmark.action_detail}
          organisation={selectedBookmark.organization}
          sources={selectedBookmark.sources}
          links={selectedBookmark.links}
          onClose={() => setSelectedBookmark(null)}
          open={selectedBookmark !== null}
        />
      )}
    </Page>
  )
}

export default Bookmarks;
