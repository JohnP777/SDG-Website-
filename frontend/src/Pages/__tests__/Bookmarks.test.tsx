import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Bookmarks from '../Bookmarks';
import * as ApiCalls from "../../Utilities/ApiCalls"
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

// Mock the API calls
jest.mock("../../Utilities/ApiCalls", () => ({
  apiCallGet: jest.fn(),
  apiCallPost: jest.fn(),
}));
  
// Mock window.location.reload
const mockReload = jest.fn();
Object.defineProperty(window, 'location', {
  value: {
    reload: mockReload,
  },
  writable: true
});

// Mock useNavigate
const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

// Helper function to render the component with router
const renderWithRouter = (component: React.ReactNode) => {
  return render(
    <MemoryRouter>
      {component}
    </MemoryRouter>
  );
};
  
describe('Bookmarks Page', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (ApiCalls.apiCallPost as jest.Mock).mockImplementation((url) => {
      if (url === 'api/auth/bookmark/set/') {
        return Promise.resolve({ statusCode: 200 });
      }
      if (url === 'api/auth/bookmark/unset/') {
        return Promise.resolve({ statusCode: 200 });
      }
    });

    (ApiCalls.apiCallGet as jest.Mock).mockImplementation((url) => {
      if (url === 'api/auth/bookmark/get/') {
        return Promise.resolve({
          bookmarks: ['action-1', 'education-2']
        });
      }
      if (url.startsWith('api/sdg-actions/retrieve')) {
        return Promise.resolve({
          id: '1',
          actions: 'Plant a tree',
          aims: 'Environmental improvement',
          action_detail: 'Planting trees to absorb carbon dioxide',
          organization: 'Green Earth Org',
          sdgs: ['1', '2', '3', '4', '5', '6', '7', '8', '9'],
          type: 'action'
        });
      }
      if (url.startsWith('api/sdg-education/retrieve')) {
        return Promise.resolve({
          id: '2',
          title: 'Climate Change Education',
          aims: 'Awareness',
          descriptions: 'Learn about climate change.',
          organization: 'UNESCO',
          sdgs_related: ['10', '11', '12', '13', '14', '15', '16', '17', '18'],
          type: 'education'
        });
      }
    });
  });

  // Snapshot test
  it("should match snapshot", async () => {
    const { asFragment } = renderWithRouter(<Bookmarks />);
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('shows bookmarks', async () => {
    renderWithRouter(<Bookmarks />);

    expect(await screen.findByText('Plant a tree')).toBeInTheDocument();
    expect(await screen.findByText('Climate Change Education')).toBeInTheDocument();
  });

  it('can search bookmarks', async () => {
    renderWithRouter(<Bookmarks />);
    
    fireEvent.change(screen.getByPlaceholderText('Search your bookmarks'), {
      target: { value: 'Climate' }
    });

    expect(await screen.findByText('Climate Change Education')).toBeInTheDocument();
    expect(await screen.queryByText('Plant a tree')).not.toBeInTheDocument();
  });

  it('opens and closes modal', async () => {
    renderWithRouter(<Bookmarks />);
    
    const bookmark = await screen.findByText('Plant a tree');
    fireEvent.click(bookmark);
    
    expect(screen.findByText('Planting trees to absorb carbon dioxide'))
    expect(screen.findByText('Organization: Green Earth Org'))

  });

  it('filters out action bookmarks when "Actions" is unchecked', async () => {
    renderWithRouter(<Bookmarks />);

    const filterButton = await screen.getByRole('button', { name: 'filter' });
    fireEvent.click(filterButton);

    const actionsCheckbox = screen.getByLabelText('Actions');
    fireEvent.click(actionsCheckbox);

    // Only education bookmark should be visible now
    expect(await screen.queryByText('Plant a tree')).not.toBeInTheDocument();
    expect(await screen.findByText('Climate Change Education')).toBeInTheDocument();
  });

  it('filters out education bookmarks when "Education" is unchecked', async () => {
    renderWithRouter(<Bookmarks />);

    const filterButton = await screen.getByRole('button', { name: 'filter' });
    fireEvent.click(filterButton);

    expect(await screen.findByText('Plant a tree')).toBeInTheDocument();
    expect(await screen.findByText('Climate Change Education')).toBeInTheDocument();

    const educationCheckbox = screen.getByLabelText('Education');
    fireEvent.click(educationCheckbox);

    // Only action bookmark should be visible now
    expect(await screen.findByText('Plant a tree')).toBeInTheDocument();
    expect(await screen.queryByText('Climate Change Education')).not.toBeInTheDocument();

  });

});
