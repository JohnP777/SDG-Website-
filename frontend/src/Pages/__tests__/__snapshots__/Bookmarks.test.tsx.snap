// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmarks Page should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-z3339j"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-u2vtxm-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: none;"
    >
      <div
        class="MuiContainer-root css-12pwiy3-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-1t0x8et-MuiToolbar-root"
        >
          <a
            class="MuiTypography-root MuiTypography-subtitle1 css-1z0p208-MuiTypography-root"
            href="/"
          >
            SDG
            <br />
            ZOO
          </a>
          <div
            class="MuiBox-root css-dq25au"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-z77o6z-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1xf0eoi-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                />
              </svg>
            </button>
          </div>
          <div
            class="MuiBox-root css-10s323d"
          >
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1oyog3l-MuiButtonBase-root-MuiButton-root"
              href="/"
              tabindex="0"
            >
              HOME
            </a>
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1oyog3l-MuiButtonBase-root-MuiButton-root"
              href="/about-us"
              tabindex="0"
            >
              ABOUT
            </a>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-769nkz-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              FEATURES
              <span
                class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="KeyboardArrowDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="MuiBox-root css-1t0e2o"
          >
            <button
              aria-label="Profile settings"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1avsmgh-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-bqh3fr-MuiAvatar-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="PersonIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-eyrs0v-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-m12f4t"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-bwxg4-MuiTypography-root"
        >
          Bookmarks
        </h2>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cxicwd-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-1e3qgpi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-zkc42u-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-qve9l1-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r6:"
              placeholder="Search your bookmarks"
              type="text"
              value=""
            />
            <button
              aria-label="filter"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="FilterAltIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M4.25 5.61C6.27 8.2 10 13 10 13v6c0 .55.45 1 1 1h2c.55 0 1-.45 1-1v-6s3.72-4.8 5.74-7.39c.51-.66.04-1.61-.79-1.61H5.04c-.83 0-1.3.95-.79 1.61"
                />
              </svg>
            </button>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-w4cd9x"
              >
                <span
                  aria-hidden="true"
                  class="notranslate"
                >
                  â€‹
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiBox-root css-1f5vcwn"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
